### ISSUE TITLE
is there anyway to set classpath folder to load extra addons or plugins Windows 

### ISSUE THREAD

**Issue by defulee** [fvarrui/JavaPackager#70]
I'm submitting a…

- [ ] bug report
- [x] feature request
- [ ] other

**Short description of the issue/suggestion:**
windows Platform do not have classpath folder to support addons or plugins;  it feature will help to build an extensible software;

**Steps to reproduce the issue/enhancement:**

1. when call launch4j-maven-plugin add classPath -> preCp parameter (which can add parameter from configuration )

**What is the expected behavior?**
add custom classpath folder to support build an extensible software;

**What is the current behavior?**
on windows platform do not have preCp parameter when call  launch4j-maven-plugin;

**Do you have outputs, screenshots, demos or samples which demonstrate the problem or enhancement?**

```
            <plugin>
                <groupId>com.akathist.maven.plugins.launch4j</groupId>
                <artifactId>launch4j-maven-plugin</artifactId>
                <executions>
                    <execution>
                        <id>...</id>
                        <phase>package</phase>
                        <goals>
                            <goal>launch4j</goal>
                        </goals>
                            <configuration>
                            ...
                            <classPath>
                                <mainClass>main.class</mainClass>
                                <preCp>classpath.folder</preCp>
                            </classPath>
                            <jre>
                                ...
                            </jre>
                            <versionInfo>
                                ...
                            </versionInfo>
                        </configuration>
                    </execution>
                </executions>
            </plugin>

```

**What is the motivation / use case for changing the behavior?**

**Please tell us about your environment:**

- JavaPackager version: 1.3.0
- OS version: windows 10
- JDK version: 11
- Build tool: 
	- [x] Maven 
	- [ ] Gradle 

**Other information** (e.g. related issues, suggestions how to fix, links for us to have context)

**Comment by fvarrui** [comment]
Hi @defulee!
In order to keep a similar behaviour for all platforms, I'd need to know how it can be done (add custom classpath folder) for Linux and MacOS.
Thanks!

**Comment by defulee** [comment]
Hi @fvarrui 

On MacOS platform we can modify info.plist ClassPath node support custom setting by new variable like `$info.classpath.customFolder`, and info.plist like:
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
	<dict>
		...
		<key>Java</key>
		<dict>
            ...
			<key>ClassPath</key>
			<string>${info.classpath.customFolder}/*:Java/${info.jarFile.name}</string>
			<key>MainClass</key>
			<string>${info.mainClass}</string>
            ...
		</dict>
        ...
	</dict>
</plist>

```

on Linux platform we can modify startup.sh.vtl use new variable `$info.classpath.customFolder` add to java command, startup.sh.vtl like :
```
#set ($vmArgs = $StringUtils.join($info.vmArgs, " "))
#!/usr/bin/env bash
# GNU/Linux startup script generated by JavaPackager plugin

...
CLASSPATH="$BINARY"
#if ($info.classpath.customFolder) 
CLASSPATH="${info.classpath.customFolder}/*:$CLASSPATH"
#end 

#if ($info.administratorRequired)
pkexec env DISPLAY=$DISPLAY XAUTHORITY=$XAUTHORITY GDKBACKEND=x11 $JAVA $JVMDefaultOptions -cp $CLASSPATH -jar $BINARY $@
#else
$JAVA $JVMDefaultOptions -cp $CLASSPATH -jar $BINARY $@
#end
exit 0
```

Thanks for your time

**Comment by fvarrui** [comment]
This is very helpful ... thanks!!! 😃

**Comment by keastrid** [comment]
Does adding the classpath to the manifest of the jar you launch not work well? We do something similar here https://github.com/keastrid/astroimagej/blob/f624efe80b0232a17af6715bdf4a8f286ca5ed2b/ij/build.gradle#L27

**Comment by defulee** [comment]
hi @keastrid 
i test on MacOs and windows platform after package MacOs app/windows exe i copy plugin jar to libs folder(classopath added in manifest manifest https://github.com/fvarrui/JavaPackager/blob/master/src/main/java/io/github/fvarrui/javapackager/maven/CreateRunnableJar.java#L57), it not work well

i do not test linux platform;

after i custom plist.info by modify Info.plist.vtl update `ClassPath` node, it works well(test on macos);

**Comment by fvarrui** [comment]
Another approach could be loading JAR files dynamically at runtime:
```java
File jarFile = new File("path/to/my.jar");
URLClassLoader child = new URLClassLoader(
        new URL[] {jarFile.toURI().toURL()},
        this.getClass().getClassLoader()
);
Class classToLoad = Class.forName("com.MyClass", true, child);
Object instanceOfMyClass = classToLoad.newInstance();
```
This would allow you to load libraries at runtime, without having to restart the JVM to extend your application with new "plugins".

**Comment by fvarrui** [comment]
I've seen [here](https://docs.oracle.com/javase/8/docs/technotes/tools/windows/java.html) that when using the `-jar` option, the .jar itself is used as the entire classpath, so the proposed script for GNU/Linux would not work.

> When you use the -jar option, the specified JAR file is the source of all user classes, and other class path settings are ignored.

**Comment by defulee** [comment]
> Another approach could be loading JAR files dynamically at runtime:
> 
> ```java
> File jarFile = new File("path/to/my.jar");
> URLClassLoader child = new URLClassLoader(
>         new URL[] {jarFile.toURI().toURL()},
>         this.getClass().getClassLoader()
> );
> Class classToLoad = Class.forName("com.MyClass", true, child);
> Object instanceOfMyClass = classToLoad.newInstance();
> ```
> 
> This would allow you to load libraries at runtime, without having to restart the JVM to extend your application with new "plugins".

thanks for your tips, I used it before; this is a good approach with hot-loading; but since I found  SPI + ServiceLoader + custom classpath is more simple approach to loader plugins(without hot-loading); I using this approach recently;

**Comment by defulee** [comment]
> I've seen [here](https://docs.oracle.com/javase/8/docs/technotes/tools/windows/java.html) that when using the `-jar` option, the .jar itself is used as the entire classpath, so the proposed script for GNU/Linux would not work.
> 
> > When you use the -jar option, the specified JAR file is the source of all user classes, and other class path settings are ignored.

yes you are right;

**Comment by fvarrui** [comment]
Hi @defulee!
Sorry for my delay! 
I've already implemented the functionality, but I have to test multiple situations and variants. I'll let you know when it's ready so you can try it.

**Comment by defulee** [comment]
> Hi @defulee!
> Sorry for my delay!
> I've already implemented the functionality, but I have to test multiple situations and variants. I'll let you know when it's ready so you can try it.

ok！thanks！this is good news!

**Comment by fvarrui** [comment]
Just as I imagined, I found a problem with a combination of properties, and maybe you could help.

There's a property called `useResourcesAsWorkingDir`, and when it's `true`, it specify that app's resources folder (on Windows is the same folder which contains exe file) must be used as the default working directory (this is the behaviour by default).

So, when I combine `classpath` and `useResourcesAsWorkingDir` as follows:
```xml
<classpath>plugins/*</classpath>
<useResourcesAsWorkingDir>false</useResourcesAsWorkingDir>
```
it looks for `plugins` in the working directory, which might not be the same as where the executable is located. This option is useful when you develop command line apps.

A first approach could be to use absolute paths in `classpath` property, but it's not a good idea ... on GNU/Linux and Mac OS it's possible to use variables when specifying `classpath`, so we can build a dynamic classpath, but I'm not sure how to do it on Windows with launch4j.

Thanks for your help!

**Comment by defulee** [comment]
> Supports executable jars and dynamic classpath resolution using environment variables and wildcards. 
> Set environment variables.

http://launch4j.sourceforge.net/  && http://launch4j.sourceforge.net/docs.html#Configuration_file 

this may help you

**Comment by fvarrui** [comment]
Another problem is the customized JRE ... if you load extra addons, maybe they need one module which it's not included, so you have to know this before add this plugin and include this module when generate the JRE with `jink` or include a full JRE r(not customized).

**Comment by keastrid** [comment]
Since users will be adding plugins themselves (or at least the possibility is available), the only safe method is to include the full JRE. 
You can't really make any guarantees about what modules are needed until usertime, unless the plugins are very limited and told beforehand what they can and cannot use.

**Comment by defulee** [comment]
@fvarrui @keastrid 
> the only safe method is to include the full JRE

yes,  that is a problem, I do not have a good idea;

**Comment by fvarrui** [comment]
I've tested on all platforms and it seems to work fine (except when combined with `useResourcesAsWorkingDir` on Mac OS)!

This change will be published in v1.4.0 asap ... until then you can try v.1.4.0-SNAPSHOT, which requires a manual installation.

**Comment by fvarrui** [comment]
v1.4.0 released to Maven Central
